contracts/contracts/harvest/BaseHarvester.sol

```diff
@@ -307,52 +437,279 @@ abstract contract BaseHarvester is Governable {
-     *      for executing this function
+     * @dev Swap a reward token for the base token on the configured
+     *      swap platform. The token must have a registered price feed
+     *      with the price provider
+     * @param _swapToken Address of the token to swap
+     * @param _rewardTo Address where to send the share of harvest rewards to
+     * @param _priceProvider Oracle to get prices of the swap token
      */
-    function _swap(address _rewardTo) internal {
-        address[] memory allStrategies = IVault(vaultAddress)
-            .getAllStrategies();
-
-        for (uint256 i = 0; i < allStrategies.length; i++) {
-            IStrategy strategy = IStrategy(allStrategies[i]);
-            address[] memory rewardTokenAddresses = strategy
-                .getRewardTokenAddresses();
-
-            for (uint256 j = 0; j < rewardTokenAddresses.length; j++) {
-                _swap(rewardTokenAddresses[j], _rewardTo);
-            }
+    function _swap(
+        address _swapToken,
+        address _rewardTo,
+        IOracle _priceProvider
+    ) internal virtual {
+        RewardTokenConfig memory tokenConfig = rewardTokenConfigs[_swapToken];
+
+        /* This will trigger a return when reward token configuration has not yet been set
+         * or we have temporarily disabled swapping of specific reward token via setting
+         * doSwapRewardToken to false.
+         */
+        if (!tokenConfig.doSwapRewardToken) {
+            return;
+        }
+
+        uint256 balance = IERC20(_swapToken).balanceOf(address(this));
+
+        if (balance == 0) {
+            return;
+        }
+
+        if (tokenConfig.liquidationLimit > 0) {
+            balance = Math.min(balance, tokenConfig.liquidationLimit);
+        }
+
+        // This'll revert if there is no price feed
+        uint256 oraclePrice = _priceProvider.price(_swapToken);
+
+        // Oracle price is 1e18
+        uint256 minExpected = (balance *
+            (1e4 - tokenConfig.allowedSlippageBps) * // max allowed slippage
+            oraclePrice).scaleBy(
+                baseTokenDecimals,
+                Helpers.getDecimals(_swapToken)
+            ) /
+            1e4 / // fix the max slippage decimal position
+            1e18; // and oracle price decimals position
+
+        // Do the swap
+        uint256 amountReceived = _doSwap(
+            tokenConfig.swapPlatform,
+            tokenConfig.swapPlatformAddr,
+            _swapToken,
+            balance,
+            minExpected
+        );
+        emit RewardTokenSwapped(
+            _swapToken,
+            baseTokenAddress,
+            tokenConfig.swapPlatform,
+            balance,
+            amountReceived
+        );
+
+        IERC20 baseToken = IERC20(baseTokenAddress);
+        uint256 baseTokenBalance = baseToken.balanceOf(address(this));
+        uint256 farmerFee = baseTokenBalance.mulTruncateScale(
```

Comment:
we should only reward the farmer on the amount swapped. ie  `amountReceived`
The farmer should not be rewarded if there is already USDT or WETH in the Harvester.

Reply:
Thanks, that's good point, will incorporate that. cc @DanielVF @sparrowDom

Reply:
Excellent point. There shouldn't be any USDT / WETH lingering on the harvester, but that doesn't make the point any less true.